stl/inc/memory

```diff
@@ -2416,26 +2437,283 @@ bool atomic_compare_exchange_weak_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty
 }
 
 template <class _Ty>
-bool atomic_compare_exchange_weak(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
+_CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_weak(
+    shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
     // atomically compare and exchange
     return _STD atomic_compare_exchange_weak_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
 template <class _Ty>
-bool atomic_compare_exchange_strong_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value,
-    memory_order, memory_order) { // atomically compare and exchange
+_CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_strong_explicit(shared_ptr<_Ty>* _Ptr,
+    shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value, memory_order, memory_order) { // atomically compare and exchange
     return _STD atomic_compare_exchange_weak_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
 template <class _Ty>
-bool atomic_compare_exchange_strong(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
+_CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_strong(
+    shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
     // atomically compare and exchange
     return _STD atomic_compare_exchange_strong_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
+#if _HAS_CXX20 && !defined(_M_CEE_PURE)
+template <class _Ty>
+struct atomic<shared_ptr<_Ty>> : _Atomic_storage<shared_ptr<_Ty>> {
+private:
+    using _Base = _Atomic_storage<shared_ptr<_Ty>>;
+
+public:
+    using value_type = shared_ptr<_Ty>;
+
+#if 1 // TRANSITION, ABI
+    static constexpr bool is_always_lock_free =
+        sizeof(shared_ptr<_Ty>) <= 8 && (sizeof(shared_ptr<_Ty>) & sizeof(shared_ptr<_Ty>) - 1) == 0;
+
+    _NODISCARD bool is_lock_free() const noexcept {
+        return is_always_lock_free;
+    }
+#else // ^^^ don't break ABI / break ABI vvv
+#if _ATOMIC_HAS_DCAS
+    static constexpr bool is_always_lock_free = sizeof(shared_ptr<_Ty>) <= 2 * sizeof(void*);
+#else // ^^^ _ATOMIC_HAS_DCAS / !_ATOMIC_HAS_DCAS vvv
+    static constexpr bool is_always_lock_free = sizeof(shared_ptr<_Ty>) <= sizeof(void*);
+#endif // _ATOMIC_HAS_DCAS
+
+    _NODISCARD bool is_lock_free() const noexcept {
+#if _ATOMIC_HAS_DCAS
+        return is_always_lock_free;
+#else // ^^^ _ATOMIC_HAS_DCAS / !_ATOMIC_HAS_DCAS vvv
+        return is_always_lock_free || (sizeof(shared_ptr<_Ty>) <= 2 * sizeof(void*) && __std_atomic_has_cmpxchg16b());
+#endif // _ATOMIC_HAS_DCAS
+    }

```

Comment:
This is a totally new type so there are no ABI break concerns.

Reply:
> This is a totally new type so there are no ABI break concerns.

It is dependent on old type. I can find a way make this use the ABI breaking section but I thought it would make a huge mess in the code.

Reply:
> I personally don't like these sorts of things. I'd prefer the non-abi break section only, with a comment detailing what you would've done in the ABI break section.

Done